# 1. Weather model

* Load the weather data from Kaggle
* Like in the previous lesson, build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?

In [7]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings(action = "ignore")

%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

from sklearn import linear_model
import statsmodels.api as sm

from sklearn.model_selection import train_test_split

In [8]:
weather = pd.read_csv("data/weatherHistory.csv")
weather.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain            9.472   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain            9.356   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain            9.378   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain            8.289   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain            8.756   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                     7.389     0.890             14.120   
1                     7.228     0.860             14.265   
2                     9.378     0.890              3.928   
3                     5.944     0.830             14.104   
4                     6.978     0.830             11.045   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                 251.000           15.826       0.000              1015.130   
1                 259.000           15.826       0.000              1015.630   
2                 204.000           14.957       0.000              1015.940   
3                 269.000           15.826       0.000              1016.410   
4                 259.000           15.826       0.000              1016.510   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [26]:
X = weather[["Humidity","Wind Speed (km/h)"]]

Y = weather["Apparent Temperature (C)"]-weather["Temperature (C)"]

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

In [28]:
model = linear_model.LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [29]:
X_test = sm.add_constant(X_test)
results = sm.OLS(y_test,X_test).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.283
Model:                            OLS   Adj. R-squared:                  0.283
Method:                 Least Squares   F-statistic:                     3810.
Date:                Thu, 17 Sep 2020   Prob (F-statistic):               0.00
Time:                        20:10:56   Log-Likelihood:                -34231.
No. Observations:               19291   AIC:                         6.847e+04
Df Residuals:                   19288   BIC:                         6.849e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4347      0.048     51.220      0.000       2.342       2.528
Humidity             -3.0331      0.054    -56.255      0.000      -3.139      -2.927
Wind Speed (km/h)    -0.1200      0.002    -77.975      0.000      -0.123      -0.117
==============================================================================
Omnibus:                      852.068   Durbin-Watson:                   2.018
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              993.141
Skew:                          -0.508   Prob(JB):                    2.20e-216
Kurtosis:                       3.450   Cond. No.                         87.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* *R-squared , Adj. R-squared are very low. These variables define the real variables %28.3.*

**Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?**

In [31]:
weather["Humidity_wind_speed"] = weather.Humidity*weather["Wind Speed (km/h)"]

In [32]:
X = weather[["Humidity","Wind Speed (km/h)","Humidity_wind_speed"]]

Y = weather["Apparent Temperature (C)"]-weather["Temperature (C)"]

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42 )

In [34]:
model = linear_model.LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [35]:
X_test = sm.add_constant(X_test)
results = sm.OLS(y_test,X_test).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.340
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     3308.
Date:                Thu, 17 Sep 2020   Prob (F-statistic):               0.00
Time:                        20:27:10   Log-Likelihood:                -33438.
No. Observations:               19291   AIC:                         6.688e+04
Df Residuals:                   19287   BIC:                         6.692e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -0.0274      0.076     -0.361      0.718      -0.176       0.121
Humidity                0.3191      0.097      3.277      0.001       0.128       0.510
Wind Speed (km/h)       0.0997      0.006     17.782      0.000       0.089       0.111
Humidity_wind_speed    -0.3113      0.008    -40.633      0.000      -0.326      -0.296
==============================================================================
Omnibus:                      968.332   Durbin-Watson:                   2.012
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1655.018
Skew:                          -0.410   Prob(JB):                         0.00
Kurtosis:                       4.177   Cond. No.                         194.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* We improved the model a little bit but It's not enough to predict the target values.

**Add visibility as additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?**

In [36]:
X = weather[["Humidity","Wind Speed (km/h)","Humidity_wind_speed","Visibility (km)"]]

Y = weather["Apparent Temperature (C)"]-weather["Temperature (C)"]

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

In [38]:
model = linear_model.LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [39]:
X_test = sm.add_constant(X_test)
results = sm.OLS(y_test,X_test).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.363
Model:                            OLS   Adj. R-squared:                  0.363
Method:                 Least Squares   F-statistic:                     2752.
Date:                Thu, 17 Sep 2020   Prob (F-statistic):               0.00
Time:                        20:35:27   Log-Likelihood:                -33086.
No. Observations:               19291   AIC:                         6.618e+04
Df Residuals:                   19286   BIC:                         6.622e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -1.2603      0.088    -14.393      0.000      -1.432      -1.089
Humidity                1.0711      0.100     10.747      0.000       0.876       1.266
Wind Speed (km/h)       0.1132      0.006     20.479      0.000       0.102       0.124
Humidity_wind_speed    -0.3317      0.008    -43.862      0.000      -0.346      -0.317
Visibility (km)         0.0669      0.002     26.774      0.000       0.062       0.072
==============================================================================
Omnibus:                     1075.191   Durbin-Watson:                   2.011
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2074.751
Skew:                          -0.410   Prob(JB):                         0.00
Kurtosis:                       4.382   Cond. No.                         247.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* There is little change in R-square.

**Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.**

* Smaller AIC and BIC scores are better for us. So, when we looked all scores we should choose last model.

# 2. House prices model

**Load the houseprices data from Kaggle.**

In [42]:
house_prices = pd.read_csv("data/houseprices.csv")
house = house_prices.select_dtypes(exclude='object')
house.isnull().sum()*100/house.shape[0]

MSSubClass       0.000
LotFrontage     17.740
LotArea          0.000
OverallQual      0.000
OverallCond      0.000
YearBuilt        0.000
YearRemodAdd     0.000
MasVnrArea       0.548
BsmtFinSF1       0.000
BsmtFinSF2       0.000
BsmtUnfSF        0.000
TotalBsmtSF      0.000
1stFlrSF         0.000
2ndFlrSF         0.000
LowQualFinSF     0.000
GrLivArea        0.000
BsmtFullBath     0.000
BsmtHalfBath     0.000
FullBath         0.000
HalfBath         0.000
BedroomAbvGr     0.000
KitchenAbvGr     0.000
TotRmsAbvGrd     0.000
Fireplaces       0.000
GarageYrBlt      5.548
GarageCars       0.000
GarageArea       0.000
WoodDeckSF       0.000
OpenPorchSF      0.000
EnclosedPorch    0.000
3SsnPorch        0.000
ScreenPorch      0.000
PoolArea         0.000
MiscVal          0.000
MoSold           0.000
YrSold           0.000
SalePrice        0.000
dtype: float64

**Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.**

In [46]:
X = house[[x for x in house.columns if x not in ["LotFrontage","SalePrice","GarageYrBlt","MasVnrArea"]]]

Y = house["SalePrice"]

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

In [48]:
model = linear_model.LinearRegression()
model.fit(X_train,y_train)

print("Coefficents : {}".format(model.coef_))
print("Intercept : {}".format(model.intercept_))

Coefficents : [-1.75548690e+02  4.09174234e-01  1.84955503e+04  3.83573710e+03
  3.37312908e+02  1.80209917e+02  1.00461555e+01 -4.61607099e+00
  3.82786747e-01  5.81287122e+00  1.31240039e+01  1.42134340e+01
  4.50422890e+00  3.18416668e+01  1.10492803e+04 -3.26518493e+02
  3.42606130e+03 -1.66531509e+03 -9.08810085e+03 -1.00971990e+04
  5.10100134e+03  4.37021557e+03  1.16074917e+04  5.58607730e-01
  2.55639049e+01 -6.62562003e+00  7.06908781e+00  3.85262374e+01
  6.55989735e+01 -3.73443286e+01 -7.15692563e-01 -2.09433750e+02
 -5.14524004e+02]
Intercept : -48411.54387291285


In [49]:
X_test = sm.add_constant(X_test)
results = sm.OLS(y_test,X_test).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.866
Model:                            OLS   Adj. R-squared:                  0.850
Method:                 Least Squares   F-statistic:                     54.04
Date:                Thu, 17 Sep 2020   Prob (F-statistic):           1.83e-95
Time:                        20:51:09   Log-Likelihood:                -3444.3
No. Observations:                 292   AIC:                             6953.
Df Residuals:                     260   BIC:                             7070.
Df Model:                          31                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const           4.32e+05   3.16e+06      0.137      0.891    -5.8e+06    6.66e+06
MSSubClass      -90.3671     60.490     -1.494      0.136    -209.479      28.745
LotArea           0.8979      0.403      2.230      0.027       0.105       1.691
OverallQual    1.392e+04   2801.257      4.970      0.000    8406.883    1.94e+04
OverallCond    7562.2941   2367.198      3.195      0.002    2900.973    1.22e+04
YearBuilt       271.1093    147.505      1.838      0.067     -19.348     561.566
YearRemodAdd      0.1949    157.715      0.001      0.999    -310.366     310.756
BsmtFinSF1       22.8507      5.732      3.987      0.000      11.564      34.137
BsmtFinSF2        5.7991      9.443      0.614      0.540     -12.796      24.394
BsmtUnfSF         0.3119      5.198      0.060      0.952      -9.924      10.548
TotalBsmtSF      28.9617      7.226      4.008      0.000      14.732      43.191
1stFlrSF         50.7316     13.410      3.783      0.000      24.325      77.138
2ndFlrSF         44.1333     12.616      3.498      0.001      19.290      68.976
LowQualFinSF    -63.4709     32.580     -1.948      0.052    -127.626       0.684
GrLivArea        31.3940     12.640      2.484      0.014       6.505      56.283
BsmtFullBath  -9932.6861   6126.607     -1.621      0.106    -2.2e+04    2131.399
BsmtHalfBath   5540.7283   8857.934      0.626      0.532   -1.19e+04     2.3e+04
FullBath       4900.2409   6315.559      0.776      0.439   -7535.915    1.73e+04
HalfBath        399.9268   6191.803      0.065      0.949   -1.18e+04    1.26e+04
BedroomAbvGr  -1.727e+04   3556.616     -4.856      0.000   -2.43e+04   -1.03e+04
KitchenAbvGr  -2.072e+04   1.28e+04     -1.624      0.106   -4.58e+04    4405.795
TotRmsAbvGrd   3021.1555   2670.327      1.131      0.259   -2237.065    8279.376
Fireplaces    -2423.8765   4062.648     -0.597      0.551   -1.04e+04    5576.006
GarageCars      1.11e+04   6271.765      1.770      0.078   -1249.130    2.35e+04
GarageArea       -5.4155     20.731     -0.261      0.794     -46.237      35.406
WoodDeckSF       55.2175     21.317      2.590      0.010      13.242      97.194
OpenPorchSF      69.5084     46.325      1.500      0.135     -21.712     160.729
EnclosedPorch   -23.5603     40.525     -0.581      0.561    -103.359      56.239
3SsnPorch      -212.2752    114.679     -1.851      0.065    -438.094      13.543
ScreenPorch      25.8486     38.777      0.667      0.506     -50.509     102.206
PoolArea         49.2722     60.560      0.814      0.417     -69.979     168.523
MiscVal          11.6799     27.792      0.420      0.675     -43.046      66.406
MoSold          912.7795    754.763      1.209      0.228    -573.447    2399.006
YrSold         -526.3356   1570.760     -0.335      0.738   -3619.367    2566.696
=================================================

* *F-statistic:	54.04*
* R-squared, adjusted R-squared are very satisfying. 
* AIC:	6953. BIC:	7070.

**Do you think your model is satisfactory? If so, why?**

I think, I found a good model because of R-squared value. We have 86 percent success on the target value. 

**In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.**

In [79]:
house["LotArea_year"] = house["LotArea"] * house ["YearBuilt"]
house["LotArea_ov_qual"] = house["LotArea"] * house["OverallQual"]
house["LotArea_ov_cond"] = house["LotArea"] * house["OverallCond"]

In [80]:
X = house[["LotArea","OverallQual","LotArea_ov_qual","OverallCond","LotArea_ov_cond","YearBuilt","LotArea_year","BsmtFinSF1","TotalBsmtSF","1stFlrSF","2ndFlrSF","LowQualFinSF","GrLivArea","BedroomAbvGr","GarageCars","WoodDeckSF"]]
Y = house["SalePrice"]

In [81]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20, random_state=42)

In [82]:
model = linear_model.LinearRegression()
model.fit(X_train,y_train)

print("Coefficents : {}".format(model.coef_))
print("Intercept : {}".format(model.intercept_))

Coefficents : [ 3.35425844e+00  2.16258713e+04 -1.54901426e-01  5.39138015e+03
  1.25307311e-02  4.04086485e+02 -9.87460614e-04  1.58351610e+01
  8.03618592e+00  2.00907087e+01  1.19322688e+01  7.55342231e+00
  3.95763998e+01 -5.26649113e+03  1.33128067e+04  2.60336400e+01]
Intercept : -897698.9021217629


In [83]:
X_test = sm.add_constant(X_test)
results = sm.OLS(y_test,X_test).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.874
Model:                            OLS   Adj. R-squared:                  0.867
Method:                 Least Squares   F-statistic:                     127.3
Date:                Fri, 18 Sep 2020   Prob (F-statistic):          2.02e-114
Time:                        09:13:40   Log-Likelihood:                -3435.2
No. Observations:                 292   AIC:                             6902.
Df Residuals:                     276   BIC:                             6961.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const             2.16e+05   4.21e+05      0.513      0.608   -6.13e+05    1.04e+06
LotArea           -94.2967     41.399     -2.278      0.024    -175.794     -12.800
OverallQual      -709.8159   4133.514     -0.172      0.864   -8847.036    7427.405
LotArea_ov_qual     1.5008      0.324      4.635      0.000       0.863       2.138
OverallCond      2369.0293   3539.845      0.669      0.504   -4599.497    9337.556
LotArea_ov_cond     0.4984      0.303      1.646      0.101      -0.098       1.095
YearBuilt        -104.2656    214.509     -0.486      0.627    -526.548     318.016
LotArea_year        0.0422      0.021      1.990      0.048       0.000       0.084
BsmtFinSF1         18.0891      5.436      3.328      0.001       7.388      28.790
TotalBsmtSF        31.3611      7.870      3.985      0.000      15.868      46.854
1stFlrSF           40.1380     11.565      3.471      0.001      17.371      62.905
2ndFlrSF           34.7724     10.701      3.249      0.001      13.705      55.839
LowQualFinSF      -36.7325     29.061     -1.264      0.207     -93.942      20.477
GrLivArea          38.1779     10.160      3.758      0.000      18.177      58.179
BedroomAbvGr    -1.096e+04   2958.275     -3.705      0.000   -1.68e+04   -5135.606
GarageCars       1.483e+04   3579.848      4.143      0.000    7785.067    2.19e+04
WoodDeckSF         34.1556     19.181      1.781      0.076      -3.604      71.916
==============================================================================
Omnibus:                      102.807   Durbin-Watson:                   2.126
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              732.694
Skew:                           1.235   Prob(JB):                    7.90e-160
Kurtosis:                      10.357   Cond. No.                     2.05e+20
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.53e-24. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

**For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?**

I think last model is the best model which we got. If we compare the r-squared values, I think 0.87 enough for our model.